### PR TITLE
Skip reviews for items that have moved to a higher level on the backend

### PR DESCRIPTION
**Pull request by mzsanford**
### 久しぶり Bug #83 

As reported in #83 and more recently [on the forum](https://community.wanikani.com/t/ios-tsurukame-native-app-with-offline-lessons-and-reviews/30635/505). Some items were moved to higher levels during a Wanikani content update, for example "久しぶり". The WK API does not accept the reviews for those but Tsurukame keeps rescheduling them since the state does not change. The results in an item showing up for review every 4 hours until you reach the level to which it has moved. In my case the estimate is some time in early 2021.

This branch filters out any assignment with a level higher than your current level. I'm not sure a settings toggle makes sense because the API does not accept it. If someone has a use case for that I can add one.

### CODE REVIEW

**Comment by turtlemaster19** [comment]
Sounds good! Nice branch name. :eyes:

File: ios/LocalCachingClient.mm [hunk 1]
```diff
@@ -657,13 +657,21 @@ - (void)maybeUpdateAvailableSubjectCounts {
...
+  TKMUser *userInfo = [self getUserInfo];
   for (TKMAssignment *assignment in assignments) {
     // Don't count assignments with invalid subjects.  This includes assignments for levels higher
-    // than the user's max level.
+    // than the user's max subscription level.
     if (![_dataLoader isValidSubjectID:assignment.subjectId]) {
       continue;
     }
 
+    // Skip assignments that are a higher level thn the user's current level. Wanikani items that have
```
Comment (davidsansome) [inline ios/LocalCachingClient.mm]: Typo: s/thn/than/

File: ios/ReviewItem.m [hunk 2]
```diff
@@ -20,13 +20,18 @@
...
-                                          dataLoader:(DataLoader *)dataLoader {
+                                          dataLoader:(DataLoader *)dataLoader
+                                  localCachingClient:(LocalCachingClient *)localCachingClient {
   NSMutableArray *ret = [NSMutableArray array];
   for (TKMAssignment *assignment in assignments) {
     if (![dataLoader isValidSubjectID:assignment.subjectId]) {
       continue;
     }
 
+    if (localCachingClient.getUserInfo.hasLevel && localCachingClient.getUserInfo.level < assignment.level) {
```
Comment (davidsansome) [inline ios/ReviewItem.m]: getUserInfo makes a database lookup so can you store it in a local variable so it only does one?

    TKMUserInfo *userInfo = [localCachingClient getUserInfo];
    if (userInfo.hasLevel && userInfo.level ...)

File: ios/LocalCachingClient.mm (hunk 1 above)
Comment (mzsanford) [reply ios/LocalCachingClient.mm]: Fixed

File: ios/ReviewItem.m (hunk 2 above)
Comment (mzsanford) [reply ios/ReviewItem.m]: I added the `getUserInfo` call before the outer loop and merged in master as well.

**Comment by turtlemaster19** [comment]
@davidsansome Can you close #83 now?

**Pull request merged**